cluster/job_once_scheduler.go

```diff
@@ -0,0 +1,236 @@
+			}
+		}
+
+		if len(keys) < keysPerPage {
+			break
+		}
+	}
+
+	return ret, nil
+}
+
+// ScheduleOnce creates a scheduled job that will run once. When the clock reaches runAt, the
+// callback will be called with key as the argument.
+//
+// If the job key already exists in the db, this will return an error. To reschedule a job, first
+// close the original then schedule it again.
+func (s *JobOnceScheduler) ScheduleOnce(key string, runAt time.Time) (*JobOnce, error) {
+	s.startedMu.RLock()
+	defer s.startedMu.RUnlock()
+	if !s.started {
+		return nil, errors.New("start the scheduler before adding jobs")
+	}
+
+	job, err := newJobOnce(s.pluginAPI, key, runAt, s.storedCallback, s.activeJobs)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create new job")
+	}
+
+	if err = job.saveMetadata(); err != nil {
+		return nil, errors.Wrap(err, "could not save job metadata")
+	}
+
+	s.runAndTrack(job)
+
+	return job, nil
+}
+
+// Close closes a job by its key. This is useful if the plugin lost the original *JobOnce, or
+// is stopping a job found in ListScheduledJobs().
+func (s *JobOnceScheduler) Close(key string) {
+	// using an anonymous function because job.Close() below needs access to the activeJobs mutex
+	job := func() *JobOnce {
+		s.activeJobs.mu.RLock()
+		defer s.activeJobs.mu.RUnlock()
+		j, ok := s.activeJobs.jobs[key]
+		if ok {
+			return j
+		}
+
+		// Job wasn't active, so no need to call Close (which shuts down the goroutine).
+		// There's a condition where another server in the cluster started the job, and the
+		// current server hasn't polled for it yet. To solve that case, delete it from the db.
+		mutex, err := NewMutex(s.pluginAPI, key)
+		if err != nil {
+			s.pluginAPI.LogError(errors.Wrap(err, "failed to create job mutex in Close for key: "+key).Error())
+		}
+		mutex.Lock()
+		defer mutex.Unlock()
+
+		_ = s.pluginAPI.KVDelete(oncePrefix + key)
+
+		return nil
+	}()
+
+	if job != nil {
+		job.Close()
+	}
+}
+
+func (s *JobOnceScheduler) scheduleNewJobsFromDB() error {
+	scheduled, err := s.ListScheduledJobs()
+	if err != nil {
+		return errors.Wrap(err, "could not read scheduled jobs from db")
+	}
+
+	for _, m := range scheduled {
+		job, err := newJobOnce(s.pluginAPI, m.Key, m.RunAt, s.storedCallback, s.activeJobs)
+		if err != nil {
+			s.pluginAPI.LogError(errors.Wrap(err, "could not create new job for key: "+m.Key).Error())
+			continue
+		}
+
+		s.runAndTrack(job)
+	}
+
+	return nil
+}
+
+func (s *JobOnceScheduler) runAndTrack(job *JobOnce) {
+	s.activeJobs.mu.Lock()
+	defer s.activeJobs.mu.Unlock()
+
+	// has this been scheduled already on this server?
+	if _, ok := s.activeJobs.jobs[job.key]; ok {
+		return
+	}
+
+	go job.run()
+
+	s.activeJobs.jobs[job.key] = job
+}
+
+// pollForNewScheduledJobs will only be started once per plugin. It doesn't need to be stopped.
+func (s *JobOnceScheduler) pollForNewScheduledJobs() {
+	for {
+		<-time.After(pollNewJobsInterval + addJitter())
```

Comment:
Will this `addJitter` add up over time in this loop? For example, if this runs 5,000 times (which will happen in ~18 days), we now have (5000 * ~50ms) = ~250s of time after the mark we want a job to run. Is there a way we can have the target time always be `pollNewJobsInterval` from the previous moment in time we ran the loop?

As in, the second time this loop runs, the time it takes to receive from `time.After` should not be affected by the previous jitter. I believe the issue takes care of itself after we have gone over 5 minutes of "lag", but we may want to be more deterministic about the time we add to the jitter.

Thinking about it some more, this loop is not intended to run every 5 minutes on the dot (5:00PM, 5:05PM), and this isn't necessarily when the jobs will actually run right? So this whole point may be null and void. Keeping this comment here for discussion.

Reply:
It will drift over time (technically in a random walk). But it doesn't matter--the polling frequency is totally arbitrary and it doesn't matter if two servers are polling at totally different times. All that matters is that they poll steadily to pick up new changes.